### PR TITLE
feat: blake3 support

### LINKED ISSUE #8650: Allow `blake3`
### Checklist

- [X] This is a bug report, not a question. Ask questions on [discuss.ipfs.io](https://discuss.ipfs.io).
- [X] I have searched on the [issue tracker](https://github.com/ipfs/go-ipfs/issues?q=is%3Aissue) for my bug.
- [X] I am running the latest [go-ipfs version](https://dist.ipfs.io/#go-ipfs) or have an issue updating.

### Installation method

ipfs-update or dist.ipfs.io

### Version

```Text
go-ipfs version: 0.11.0
Repo version: 11
System version: amd64/windows
Golang version: go1.16.12
```

### Config

```json
irrelevant
```

### Description

This command gives an error message:

```
ipfs add --hash=blake3 some.file
Error: potentially insecure hash functions not allowed
```

While using `sha1` produces a valid CIDv1

```
ipfs add --hash=sha1 some.file
```

`sha1` is known to be broken, `blake3` is not. So it would make sense to either remove the security warning for `blake3` or add one for `sha1`.

### PR DESCRIPTION
**Pull request by laudiacay**
Accepting the relevant blake3 PRs on go-multihash, go-cid, and go-verifcid will make this one ready to go- it's just version bumps that make blake3 work on ipfs. all validation happens in called libraries, the ipfs code more or less doesn't care what hash function multihash is using as far as i can tell.

Closes #8650

### CODE REVIEW

**Comment by Jorropo** [comment]
pls run `go mod tidy`

**Comment by laudiacay** [comment]
Closes #8650.

prereqs:
PR 1: https://github.com/multiformats/go-multihash/pull/158
PR 2: https://github.com/ipfs/go-cid/pull/140
PR 3: https://github.com/ipfs/go-verifcid/pull/15

**Comment by laudiacay** [comment]
> pls run `go mod tidy`

won't run until the other version bump PRs are live- will do before merging.

**Comment by Jorropo** [comment]
> won't run until the other version bump PRs are live- will do before merging.

Yes mb, what we do most of the time is use replace rules or `go get repo@PRCommit` (`PRCommit` is literally just the head commit of your PR) so we can test them together and see the green CI passing.

**Review by lidel — CHANGES_REQUESTED** [review]
### We need to be extra careful here

IIUC this PR introduces `blake3` with implicit length of 256 bits (32 bytes), however it is very likely at some point in the future we will be asked to add support for  longer versions. 

This means the bar here is higher than it would be for existing functions: we not only need to guard the behavior at the implicit default length proposed in this PR, but also have a plan for supporting alternative lengths under the same name.

### UX challenges

Custom length can be passed to `block put`  via `--mhlen`. afaik it was intended for digest truncation in contexts where data from third party systems, but in the context of blake3 could be used for producing longer digests. 

Other commands like `ipfs dag put` and `ipfs add` do not have `--hash-len`.  We need to decide if we want to add them as part of this PR, or leave that for future PR, when a need for longer blake3 is raised by our users.  I think leaving it for future is fine.

###  TODO

Things that needs to happen before we can merge this PR:

- [ ] decide if this PR will also add `--hash-len` to `add` and `dag put` commands  (only for blake3)
- [ ] add sharness tests listed  in https://github.com/ipfs/go-ipfs/issues/8650#issuecomment-1172552991

**Comment by laudiacay** [comment]
@lidel some relevant discussion here: https://github.com/ipfs/go-verifcid/pull/15#discussion_r915263122

also discussed with @aschmahmann in DM about this.

tl;dr: supporting multiple lengths in any manner except literally having a bunch of different multihash codes for them is going to require a major refactor of the multihash library and interface. i don't have time to implement that right now, it's gonna be really painful... but 32-bytes is standard and gets the work that's needed for my usecase (and probably for many others') done. 

right now i have it throwing an error on trying to parse blake3 hashes that are not 32 bytes long, no way to create them without somehow subverting the multihash interface, and tests for that behavior in the verifcid library...

assuming I get tests passing... how do people feel about merging it like this and worrying about the interface whenever someone wants to go through and actually refactor multihash because they need something beyond 32 bytes?

**Comment by laudiacay** [comment]
i think we're also pending some sharness. adding now...

**Comment by Jorropo** [comment]
replaced #9187

**Comment by Jorropo** [comment]
@laudiacay thx for your patchers, this has been merged in #9187 :tada:
